File: lib/hooks/transactionMappers.ts
// lib/hooks/transactionMappers.ts
import { useCallback } from 'react';
import type { Transaction } from '../../types';
import { Recurrence } from '../../types';
import { SYSTEM_CATEGORIES } from '../../constants';

// Valid recurrence values that must match the database CHECK constraint
const VALID_RECURRENCES = [
  Recurrence.ONE_TIME,
  Recurrence.BIWEEKLY,
  Recurrence.MONTHLY,
];

const normalizeBudgetName = (value: string) => value.trim().toLowerCase();

const toLocalIsoDay = (value: Date) => {
  const year = value.getFullYear();
  const month = String(value.getMonth() + 1).padStart(2, '0');
  const day = String(value.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
};

const toIsoDay = (value: unknown): string => {
  if (typeof value === 'string' && /^\d{4}-\d{2}-\d{2}/.test(value)) {
    return value.slice(0, 10);
  }

  const parsed = new Date(String(value));
  if (Number.isNaN(parsed.getTime())) return '';
  return parsed.toISOString().slice(0, 10);
};

export const shouldSolidifyProjectedTransaction = (
  isProjected: boolean,
  transactionDate: unknown,
  now: Date = new Date(),
) => {
  if (!isProjected) return false;
  const txIsoDay = toIsoDay(transactionDate);
  if (!txIsoDay) return false;
  return txIsoDay <= toLocalIsoDay(now);
};

const systemCategoryIdByName = new Map(
  SYSTEM_CATEGORIES.map(category => [normalizeBudgetName(category.name), category.id]),
);

export const resolveBudgetNameForInsert = (
  budgetId: string | null,
  budgets: { id: string; name: string }[] = [],
): string => {
  if (!budgetId) {
    throw new Error(`Transaction must have a valid budget_id (category_id). Got: ${budgetId}`);
  }

  const directIdMatch = budgets.find(b => b.id === budgetId);
  if (directIdMatch?.name) return directIdMatch.name;

  const normalizedInput = normalizeBudgetName(budgetId);
  const nameMatch = budgets.find(b => normalizeBudgetName(b.name) === normalizedInput);
  if (nameMatch?.name) return nameMatch.name;

  if (budgetId.startsWith('budget:')) {
    const fromPrefixedId = budgetId.slice('budget:'.length).replace(/-/g, ' ');
    const prefixedMatch = budgets.find(
      b => normalizeBudgetName(b.name) === normalizeBudgetName(fromPrefixedId),
    );
    if (prefixedMatch?.name) return prefixedMatch.name;
  }

  throw new Error(
    `Cannot map budget_id "${budgetId}" to a valid budget name for transactions.budget`,
  );
};

export const resolveBudgetIdFromRow = (row: any): string | null => {
  const budgetRaw = row.Budget || row.budget;
  if (budgetRaw) {
    const budgetName = String(budgetRaw);
    const normalizedName = normalizeBudgetName(budgetName);
    const systemId = systemCategoryIdByName.get(normalizedName);
    if (systemId) return systemId;

    return `budget:${normalizedName.replace(/\s+/g, '-')}`;
  }

  return null;
};

// Build the object Supabase expects — only columns that exist in the table
export const toSupabaseTransaction = (
  tx: Transaction,
  budgets: { id: string; name: string }[] = [],
) => {
  // Extract the YYYY-MM-DD portion directly from the date string to avoid
  // timezone-related date shifts that occur when round-tripping through the
  // Date constructor (e.g. "2025-03-01T00:00:00Z" parsed in UTC-8 becomes
  // Feb 28 locally).
  const dateStr = tx.date.slice(0, 10);

  // Validate required fields
  if (!tx.budget_id) {
    throw new Error(`Transaction must have a valid budget_id (category_id). Got: ${tx.budget_id}`);
  }

  // Validate and set recurrence value
  let recurrence: string = Recurrence.ONE_TIME;
  if (tx.recurrence) {
    if (VALID_RECURRENCES.includes(tx.recurrence as Recurrence)) {
      recurrence = tx.recurrence;
    } else {
      console.warn(`Invalid recurrence value "${tx.recurrence}", defaulting to "${Recurrence.ONE_TIME}"`);
    }
  }

  const budgetName = resolveBudgetNameForInsert(tx.budget_id, budgets);

  const row: Record<string, any> = {
    id: tx.id,
    user_id: tx.user_id,
    vendor: tx.vendor,
    amount: Number(tx.amount),
    date: dateStr,
    is_projected: tx.is_projected ?? false,
    // Current schema columns (public.transactions)
    // type enum only has 'Manual' and 'Automatic' — map AI/Automatic labels to 'Automatic'
    budget: budgetName,
    type: tx.label === 'Automatic' ? 'Automatic' : 'Manual',
    recur: recurrence,
  };

  return row;
};

export const useToSupabaseTransaction = (budgets: { id: string; name: string }[] = []) =>
  useCallback((tx: Transaction) => toSupabaseTransaction(tx, budgets), [budgets]);

// Convert Supabase transaction to app format
export const useFromSupabaseTransaction = () =>
  useCallback((row: any): Transaction => {
    // Validate recurrence value from database (supports recur or recurrence)
    let recurrence: Recurrence = Recurrence.ONE_TIME;
    const recurrenceRaw = row.recur || row.recurrence;
    if (recurrenceRaw) {
      if (VALID_RECURRENCES.includes(recurrenceRaw as Recurrence)) {
        recurrence = recurrenceRaw as Recurrence;
      } else {
        console.warn(`Invalid recurrence value "${recurrenceRaw}" from database, using "${Recurrence.ONE_TIME}"`);
      }
    }

    const shouldSolidify = shouldSolidifyProjectedTransaction(row.is_projected, row.date);

    return {
      id: row.id,
      user_id: row.user_id,
      vendor: row.vendor,
      amount: parseFloat(row.amount),
      // Keep date as a YYYY-MM-DD string (with a noon-UTC timestamp appended so
      // that slicing to 10 chars always yields the correct calendar date regardless
      // of the user's timezone).
      date: typeof row.date === 'string' && /^\d{4}-\d{2}-\d{2}$/.test(row.date)
        ? row.date + 'T12:00:00.000Z'
        : new Date(row.date).toISOString(),
      budget_id: resolveBudgetIdFromRow(row),
      recurrence: recurrence,
      // Map DB type enum ('Automatic'|'Manual') to app-level label.
      label: row.type === 'Automatic' ? 'Automatic' : 'Manual',
      is_projected: shouldSolidify ? false : row.is_projected,
      is_income: row.is_income === true,
      caught_cleared: row.caught_cleared === true,
      userName: row.user_name || '',
      created_at: row.created_at,
    };
  }, []);

File: android-kotlin/app/src/main/java/com/covault/app/data/remote/TransactionMappers.kt
package com.covault.app.data.remote

import com.covault.app.data.model.BudgetCategory
import com.covault.app.data.model.Recurrence
import com.covault.app.data.model.SubscriptionStatus
import com.covault.app.data.model.SystemCategories
import com.covault.app.data.model.Transaction
import com.covault.app.data.model.TransactionLabel
import com.covault.app.data.model.TransactionSource
import com.covault.app.data.model.User
import com.covault.app.data.remote.dto.BudgetRow
import com.covault.app.data.remote.dto.SettingsRow
import com.covault.app.data.remote.dto.TransactionRow
import com.covault.app.domain.DateUtils.toLocalIsoDay
import java.time.LocalDate
import java.time.format.DateTimeFormatter

/**
 * Conversion between Supabase row DTOs and the domain model the UI uses.
 * Direct port of `lib/hooks/transactionMappers.ts` and the relevant parts
 * of `lib/useAuthState.ts` / `lib/useUserData.ts`.
 *
 * All date string handling uses [DateUtils] to stay in the local calendar.
 */
object TransactionMappers {

    private val validRecurrences = setOf("One-time", "Biweekly", "Monthly")

    /**
     * True when a projected transaction has reached or passed today and
     * should be treated as a real (non-projected) row. Mirrors
     * `shouldSolidifyProjectedTransaction` in the React code.
     */
    fun shouldSolidifyProjectedTransaction(
        isProjected: Boolean,
        transactionDate: String,
        now: LocalDate = LocalDate.now(),
    ): Boolean {
        if (!isProjected) return false
        val txDay = parseIsoDay(transactionDate) ?: return false
        return !txDay.isAfter(now)
    }

    private fun parseIsoDay(value: String): LocalDate? {
        if (value.length >= 10) {
            return runCatching { LocalDate.parse(value.substring(0, 10)) }.getOrNull()
        }
        return null
    }

    /**
     * Resolve a transaction's `budget_id` (system UUID) from a Supabase
     * row's `budget` enum value. Returns null if the row has no budget
     * field — the caller is expected to reject the row in that case
     * (the React app's `toSupabaseTransaction` throws if budget_id is
     * missing on the outbound side).
     */
    fun resolveBudgetIdFromRow(row: TransactionRow): String? {
        val name = row.budget.takeIf { it.isNotBlank() } ?: return null
        val normalized = name.trim().lowercase()
        return SystemCategories.idForName(normalized)
            ?: "budget:${normalized.replace("\\s+".toRegex(), "-")}"
    }

    /**
     * Inverse of [resolveBudgetIdFromRow]. Throws if the budget_id cannot
     * be mapped to a valid `Budgets` enum value.
     */
    fun resolveBudgetNameForInsert(
        budgetId: String?,
        budgets: List<BudgetCategory> = emptyList(),
    ): String {
        require(!budgetId.isNullOrBlank()) {
            "Transaction must have a valid budget_id (category_id). Got: $budgetId"
        }
        budgets.firstOrNull { it.id == budgetId }?.name?.let { return it }

        val normalized = budgetId.trim().lowercase()
        budgets.firstOrNull { it.name.trim().lowercase() == normalized }?.name?.let { return it }

        if (budgetId.startsWith("budget:")) {
            val fromPrefixed = budgetId.removePrefix("budget:").replace("-", " ")
            budgets.firstOrNull { it.name.trim().lowercase() == fromPrefixed.trim().lowercase() }
                ?.name?.let { return it }
        }

        throw IllegalArgumentException(
            "Cannot map budget_id \"$budgetId\" to a valid budget name for transactions.budget"
        )
    }

    /**
     * Build the object Supabase expects for an insert/update. Only includes
     * columns that exist in the live `public.transactions` table.
     */
    fun toSupabaseRow(
        tx: Transaction,
        budgets: List<BudgetCategory> = emptyList(),
    ): Map<String, Any?> {
        val dateStr = tx.date.substring(0, 10)
        require(tx.budgetId != null) {
            "Transaction must have a valid budget_id (category_id). Got: ${tx.budgetId}"
        }

        val recurrence: String = if (tx.recurrence.dbValue in validRecurrences) {
            tx.recurrence.dbValue
        } else {
            Recurrence.ONE_TIME.dbValue
        }
        val budgetName = resolveBudgetNameForInsert(tx.budgetId, budgets)

        val row = mutableMapOf<String, Any?>(
            "id" to tx.id,
            "user_id" to tx.userId,
            "vendor" to tx.vendor,
            "amount" to tx.amount,
            "date" to dateStr,
            "is_projected" to (tx.isProjected),
            "budget" to budgetName,
            "type" to if (tx.label == TransactionLabel.AUTOMATIC) "Automatic" else "Manual",
            "recur" to recurrence,
        )
        tx.source?.let { row["source"] = it.name.lowercase() }
        return row
    }

    /**
     * Convert a Supabase row to the domain [Transaction]. The date string
     * is normalized to `YYYY-MM-DDT12:00:00.000Z` so that slicing to
     * 10 chars always yields the correct calendar date in every timezone
     * (matches the React `useFromSupabaseTransaction` behavior).
     */
    fun fromSupabaseRow(row: TransactionRow): Transaction {
        val recurrence = if (row.recur in validRecurrences) {
            Recurrence.fromDbValue(row.recur)
        } else {
            Recurrence.ONE_TIME
        }
        val shouldSolidify = shouldSolidifyProjectedTransaction(
            isProjected = row.isProjected,
            transactionDate = row.date,
        )

        val dateString: String = if (row.date.matches(Regex("^\\d{4}-\\d{2}-\\d{2}$"))) {
            row.date + "T12:00:00.000Z"
        } else {
            runCatching { LocalDate.parse(row.date.substring(0, 10))
                .atStartOfDay(java.time.ZoneOffset.UTC)
                .toInstant()
                .toString() }
                .getOrDefault(row.date)
        }

        return Transaction(
            id = row.id,
            userId = row.userId,
            vendor = row.vendor,
            amount = row.amount,
            date = dateString,
            budgetId = resolveBudgetIdFromRow(row),
            recurrence = recurrence,
            label = if (row.type == "Automatic") TransactionLabel.AUTOMATIC else TransactionLabel.MANUAL,
            isProjected = if (shouldSolidify) false else row.isProjected,
            isIncome = row.isIncome == true,
            caughtCleared = row.caughtCleared == true,
            userName = row.userName ?: "",
            createdAt = row.createdAt.orEmpty(),
            source = runCatching { TransactionSource.valueOf(row.source?.uppercase().orEmpty()) }
                .getOrNull(),
        )
    }

    // ------------------------------------------------------------------------
    // Settings / User mappers
    // ------------------------------------------------------------------------

    fun userFromSettings(row: SettingsRow): User = User(
        id = row.userId,
        name = row.name,
        email = row.email,
        partnerId = row.partnerId,
        partnerEmail = row.partnerEmail,
        partnerName = row.partnerName,
        hasJointAccounts = (row.partnerId != null),
        budgetingSolo = row.budgetingSolo ?: true,
        monthlyIncome = row.monthlyIncome ?: 0.0,
        trialStartedAt = row.trialStartedAt,
        trialEndsAt = row.trialEndsAt,
        trialConsumed = row.trialConsumed == true,
        subscriptionStatus = when (row.subscriptionStatus) {
            "active" -> SubscriptionStatus.ACTIVE
            "expired" -> SubscriptionStatus.EXPIRED
            else -> SubscriptionStatus.NONE
        },
    )

    fun budgetCategoryFromRow(row: BudgetRow): BudgetCategory? {
        val id = SystemCategories.idForName(row.budget) ?: return null
        return BudgetCategory(
            id = id,
            name = SystemCategories.nameForId(id) ?: return null,
            totalLimit = row.amount,
        )
    }
}
